### PR TITLE
Reorganize and make things more typed, as well as improve the CLI

### PR DESCRIPTION
**Pull request by orowith2os**
I also improved the parsing of the websocket URL, so now it returns actual errors. A good thing to do after this would be to return the errors as plain Strings too, that way you don't get nasty debug-esque results.

### CODE REVIEW

**Comment by orowith2os** [comment]
I'm not sure if type-safe is the right word here - but it's definitely more typed than it was before.

**Comment by grigio** [comment]
Good reorganization, did you try that all subcommands work? If you can pass clippy `cargo fmt -- --check` I'll merge it.

**Comment by orowith2os** [comment]
Ran `cargo fmt` and fixed some nits from `cargo clippy`.  Also ran `cargo fmt -- --check` and it should pass now, at least it does locally. 

All subcommands *should* be working fine now, with better errors on the CLI now (e.g. command recs when you type something wrong, and it lists out all the different choices). Everything shows up properly on my end, and testing some commands, they all work.

I will note that the user is now forced to input the URL for the websocket, but I don't see that as a big problem so long as the README (should I add that to the CLI as an error too?) tells them what to do.

**Comment by grigio** [comment]
> I will note that the user is now forced to input the URL for the websocket, but I don't see that as a big problem so long as the README (should I add that to the CLI as an error too?) tells them what to do.

Yes, I'd like to keep the thing simple and not overengineered a default `obsws` url  and parameter `--obsws` for who wants to override it. Feel free to re-add the `obs-cmd.toml` support if you need it

**Comment by orowith2os** [comment]
No need, I'll just `match` on `cli.websocket` and provide default values if `None`, as well as wrap `websocket` as `Option<ObsWebsocket>`.

File: src/command.rs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+#[derive(Parser)]
+#[clap(author, version, about, long_about = None)]
+pub struct Cli {
+    #[clap(short, long)]
+    pub websocket: ObsWebsocket,
+    #[clap(subcommand)]
+    pub command: Commands,
+}
```
[the remark this answers is not in the document]
Comment (orowith2os) [reply src/command.rs:86]: I'll use it with the websocket argument to show you what I mean.

**Comment by grigio** [comment]
Can you check that the parameter `--obsws` continue to work also in the explicit way?

```
obs-cmd git:(master) ./target/debug/obs-cmd --obsws obsws://localhost:4455/secret info
Version: Version { obs_version: Version { major: 30, minor: 0, patch: 0 }, obs_web_socket_version: Version { major: 5, minor: 3, patch: 3 }, rpc_version: 1, available_requests: ["OpenSourceProjector", "OpenInputInteractDialog", "OpenInputFiltersDialog", "OpenInputPropertiesDialog", "OffsetMediaInputCursor", "GetMediaInputStatus", "ResumeRecord", "SetInputAudioSyncOffset",
```

I tested your branch but i gave me issue

```
obs-cmd git:(cleanups) ./target/debug/obs-cmd --websocket obsws://localhost:4455/secret info
Error: Handshake(ConnectionClosed(Some(CloseDetails { code: Library(4009), reason: "Authentication failed." })))

obs-cmd git:(cleanups) ./target/debug/obs-cmd --obsws obsws://localhost:4455/secret info
error: unexpected argument '--obsws' found

Usage: obs-cmd [OPTIONS] <COMMAND>


```

**Comment by orowith2os** [comment]
I renamed the argument to `--websocket`, which should get the point across better.

**Comment by orowith2os** [comment]
After mulling over the changes a bit more, I'm not seeing any regressions or major bugs; everything necessary should be handled. We validate the websocket URL as best we can (so it's valid when passed to the API) and the CLI should still be compatible aside from the `--websocket` flag change.

There are a few more possible cleanups in `main.rs`, but I'm going to save those for later.

**Comment by grigio** [comment]
image

I'd like to merge this, but it should behave like in `master` branch, with the default url and also with the explicit url

**Comment by orowith2os** [comment]
It not working there seems like a bug somewhere. I'll add some debug logging and take a peek.

**Comment by orowith2os** [comment]
Found the problem.

```rust
⬢ [crystalbox2:2023-07-29] ❯ cargo run -- -w obsws://localhost:4455/secret info
    Finished dev [unoptimized + debuginfo] target(s) in 0.05s
     Running `target/debug/obs-cmd -w 'obsws://localhost:4455/secret' info`
[src/main.rs:11] cli.websocket.clone() = Some(
    ObsWebsocket {
        hostname: "localhost",
        port: 4455,
        password: Some(
            "/secret",
        ),
    },
)
Error: Connect(Io(Os { code: 111, kind: ConnectionRefused, message: "Connection refused" }))
```

**Comment by orowith2os** [comment]
Try now.

**Comment by grigio** [comment]
Nope

it seems the url is not split correctly and maybe the default password is `secret` and not `/secret`. It shuould behave like https://github.com/grigio/obs-cmd/blob/master/src/main.rs#L42

```
➜  obs-cmd git:(cleanups) cargo run -- -w obsws://localhost:4455/secret info 
    Finished dev [unoptimized + debuginfo] target(s) in 0.37s
     Running `target/debug/obs-cmd -w 'obsws://localhost:4455/secret' info`
Error: Connect(Io(Custom { kind: Uncategorized, error: "failed to lookup address information: Name or service not known" }))
```

**Comment by orowith2os** [comment]
Eugh. I keep messing up the logic. Looks like I split the hostname, not the password.
```rust
[src/main.rs:11] cli.websocket.clone() = Some(
    ObsWebsocket {
        hostname: "ocalhost",
        port: 4455,
        password: Some(
            "/secret",
        ),
    },
)
```

Idea still stands, I just did it wrong.

**Review by grigio — APPROVED** [review]
It seems OK now

**Comment by orowith2os** [comment]
If there's anything else regarding this tool you'd like me to work on, poke me - I've been enjoying it, and if you're willing to listen to how it all works, I'd be happy to help teach you Rust.

**Pull request merged**